src/components/tl/mlx5/mcast/tl_mlx5_mcast_allgather.c

```diff
@@ -0,0 +1,424 @@
+        status = UCC_ERR_NOT_SUPPORTED;
+        goto failed;
+    }
+
+    req->last_pkt_len = req->length - (req->num_packets - 1)*comm->max_per_packet;
+
+    ucc_assert(req->last_pkt_len > 0 && req->last_pkt_len <= comm->max_per_packet);
+
+    if (req->proto == MCAST_PROTO_ZCOPY) {
+        /* register the send buffer */
+       status = ucc_tl_mlx5_mcast_mem_register(comm->ctx, req->ptr, req->length, &reg);
+       if (UCC_OK != status) {
+           goto failed;
+       }
+       req->rreg = reg;
+       req->mr   = reg->mr;
+    }
+
+    if (comm->one_sided.reliability_enabled) {
+        req->one_sided_reliability_scheme = (req->length <
+                comm->one_sided.reliability_scheme_msg_threshold) ?
+                ONE_SIDED_ASYNCHRONOUS_PROTO : ONE_SIDED_SYNCHRONOUS_PROTO;
+    } else {
+        req->one_sided_reliability_scheme = ONE_SIDED_NO_RELIABILITY;
+    }
+
+    req->ag_counter = comm->allgather_comm.coll_counter;
+    req->to_send    = req->num_packets;
+    req->to_recv    = comm->commsize * req->num_packets;
+
+    comm->allgather_comm.coll_counter++;
+
+    task->coll_mcast.req_handle = req;
+    coll_task->status           = UCC_INPROGRESS;
+    return ucc_progress_queue_enqueue(UCC_TL_CORE_CTX(mlx5_team)->pq, &task->super);
+
+failed:
+    tl_warn(UCC_TASK_LIB(task), "mcast start allgather failed:%d", status);
+    if (req) {
+        ucc_free(req);
+    }
+    coll_task->status = status;
+    return ucc_task_complete(coll_task);
+}
+
+void ucc_tl_mlx5_mcast_allgather_progress(ucc_coll_task_t *coll_task)
+{
+    ucc_tl_mlx5_task_t           *task = ucc_derived_of(coll_task, ucc_tl_mlx5_task_t);
+    ucc_tl_mlx5_mcast_coll_req_t *req  = task->coll_mcast.req_handle;
+    ucc_status_t                  status;
+
+    if (req != NULL) {
+        if (req->ag_counter != req->comm->allgather_comm.under_progress_counter) {
+            /* it is not this task's turn for progress */
+            ucc_assert(req->comm->allgather_comm.under_progress_counter < req->ag_counter);
+            return;
+        }
+
+        status = ucc_tl_mlx5_mcast_test_allgather(task->coll_mcast.req_handle);
+        if (UCC_INPROGRESS == status) {
+            return;
+        } else if (UCC_OK == status) {
+            coll_task->status = UCC_OK;
+            req->comm->allgather_comm.under_progress_counter++;
+            ucc_free(req);
+            task->coll_mcast.req_handle = NULL;
+        } else {
+            tl_error(UCC_TASK_LIB(task), "progress mcast allgather failed:%d", status);
+            coll_task->status = status;
+            if (req->rreg) {
+                ucc_tl_mlx5_mcast_mem_deregister(req->comm->ctx, req->rreg);
+                req->rreg = NULL;
+            }
+            if (req->recv_rreg) {
+                ucc_tl_mlx5_mcast_mem_deregister(req->comm->ctx, req->recv_rreg);
+                req->recv_rreg = NULL;
+            }
+            ucc_free(req);
+            ucc_task_complete(coll_task);
```

Comment:
ucc_task_complete will be called from ucc progress queue, need to set error status